### PR TITLE
[wasm] crypto deadlock CI test

### PR DESCRIPTION
**Pull request by pavelsavara**
just testing

### CODE REVIEW

**Comment by pavelsavara** [comment]
```
warn: MAIN1: Probably deadlock during read_response with state=0Error
    at LibraryChannel.wait_for_state (http://127.0.0.1:60672/dotnet.js:3674:103)
    at LibraryChannel.read_response (http://127.0.0.1:60672/dotnet.js:3638:36)
    at LibraryChannel.send_msg (http://127.0.0.1:60672/dotnet.js:3562:29)
    at _send_msg_worker (http://127.0.0.1:60672/dotnet.js:3499:55)
    at Object.dotnet_browser_encrypt_decrypt (http://127.0.0.1:60672/dotnet.js:3438:24)
    at _dotnet_browser_encrypt_decrypt (http://127.0.0.1:60672/dotnet.js:15377:75)
    at SystemCryptoNativeBrowser_EncryptDecrypt (SystemCryptoNativeBrowser_EncryptDecrypt (http://127.0.0.1:60672/dotnet.wasm:wasm-function[17135]:0x2e71f7))
    at wasm_invoke_iiiiiiiiii (wasm_invoke_iiiiiiiiii (http://127.0.0.1:60672/dotnet.wasm:wasm-function[193]:0x9b1a))
    at ves_pinvoke_method (ves_pinvoke_method (http://127.0.0.1:60672/dotnet.wasm:wasm-function[290]:0x1cbef))
    at interp_exec_method (interp_exec_method (http://127.0.0.1:60672/dotnet.wasm:wasm-function[238]:0xc28f))
```

**Comment by pavelsavara** [comment]
I added tracing to `wait_for_state` in main thread

```
MAIN1: Probably deadlock during read_response with state=1
    at LibraryChannel.wait_for_state (http://127.0.0.1:52184/dotnet.js:3684:103)
    at LibraryChannel.read_response (http://127.0.0.1:52184/dotnet.js:3641:36)
    at LibraryChannel.send_msg (http://127.0.0.1:52184/dotnet.js:3565:29)
    at _send_msg_worker (http://127.0.0.1:52184/dotnet.js:3502:55)
    at Object.dotnet_browser_encrypt_decrypt (http://127.0.0.1:52184/dotnet.js:3441:24)
    at _dotnet_browser_encrypt_decrypt (http://127.0.0.1:52184/dotnet.js:15380:75)
    at SystemCryptoNativeBrowser_EncryptDecrypt (SystemCryptoNativeBrowser_EncryptDecrypt (http://127.0.0.1:52184
```
and to `_wait` on worker thread.
```
info: [crypto-worker] wait failed with not-equal  actual 1 when waiting for 0
    at ChannelWorker._wait (http://127.0.0.1:52184/dotnet-crypto-worker.js:278:112)
    at ChannelWorker.run_message_loop (http://127.0.0.1:52184/dotnet-crypto-worker.js:118:30)
```

The state is `STATE_IDLE` but the worker is waiting for `STATE_REQ`. The main is waiting for `STATE_RESP || STATE_RESP_P`.

cc @eerhardt @radical

**Comment by pavelsavara** [comment]
Repro:
1) 
`dotnet build /p:TargetOS=Browser /p:TargetArchitecture=wasm /p:Configuration=Debug /t:Test src\libraries\System.Security.Cryptography\tests /p:Scenario=WasmTestOnBrowser /p:UseSubtleCryptoForTests=true /p:WasmTestAppArgs="-class System.Security.Cryptography.Encryption.Aes.Tests.AesContractTests"`

2) 
run `a.cmd 1` in console 1
run `a.cmd 2` in console 2
run `a.cmd n` in console n

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 1]
```diff
@@ -269,19 +263,17 @@ class LibraryChannel {
...
-                this.release_lock();
-            }
-
+            });
+            // Notify webworker
             Atomics.notify(this.comm, this.STATE_IDX);
 
             // The send message is complete.
-            if (state === this.STATE_REQ)
+            if (state === this.STATE_REQ) {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/crypto-worker.ts:272]: could state become something else than `STATE_REQ` or `STATE_REQ_P` ?

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 2]
```diff
@@ -371,7 +378,7 @@ class LibraryChannel {
         }
     }
 
-    private release_lock() {
+    private _release_lock() {
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:381]: The function is marked `private` - does it need to have an underscore prefix?

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 3]
```diff
@@ -358,7 +355,17 @@ class LibraryChannel {
...
+    private _using_lock_for_nonblocking_cb(cb: Function) {
+        try {
+            this._acquire_lock_with_spin_wait();
+            return cb();
+        } finally {
+            this._release_lock();
+        }
+    }
+
+    private _acquire_lock_with_spin_wait() {
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:368]: Is it common to put implementation details in the name of the method? "With spin wait" seems more appropriate for a comment than apart of the method name.

Similar comment for `_using_lock_for_nonblocking_cb`.

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 4]
```diff
@@ -335,20 +333,19 @@ class LibraryChannel {
...
+                const state = Atomics.load(this.comm, this.STATE_IDX);
+                if (state == this.STATE_REQ_FAILED)
+                    throw new OperationFailedError(`Worker failed during ${msg} with state=${state}`);
 
-            if (is_ready(state))
-                return state;
+                if (is_ready(state))
+                    return true;
+            });
+            if (done) return;
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:348]: (nit) it is much more readable to put the body of the `if` on a new line.

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 5]
```diff
@@ -299,33 +291,39 @@ class LibraryChannel {
...
+            this.wait_for_state_change_to(state => state == this.STATE_RESP || state == this.STATE_RESP_P, "read_response");
+            let state;
+            const done = this._using_lock_for_nonblocking_cb(() => {
                 const size_to_read = Atomics.load(this.comm, this.MSG_SIZE_IDX);
 
                 // Append the latest part of the message.
                 response += this.read_from_msg(0, size_to_read);
 
                 // The response is complete.
+                state = Atomics.load(this.comm, this.STATE_IDX);
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:303]: Why do we need this extra read of the state? Once the state is `STATE_RESP || STATE_RESP_P`, the only thing that can (should?) change the state is the main thread (i.e. this code).

File: src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 import { setup_proxy_console } from "../debug";
 import type { InitCryptoMessageData } from "../crypto-worker";
-import type { MonoConfig } from "../types";
+// import type { MonoConfig } from "../types";
```
Comment (eerhardt) [inline src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:6]: Can we delete commented out code?

Comment (pavelsavara) [reply src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:6]: Sorry Eric, this is just PR to be able to run CI tests. I will give you link to actual clean fix PR soon

**Pull request closed without merging**